[4.2] Fix fatal error when update to Joomla 4.2

joomdonation opened the pull request:
Pull Request for Issue # .

### Summary of Changes
In Joomla 4.2, we convert some plugins to new structure, so the namespace map will need to be re-generated before avoid Joomla calls com_joomlaupdate to finalize upgrade process to avoid fatal error.

### Testing Instructions
1. Install Joomla 4.1.
2. Try to Update that 4.1 website to Joomla 4.2 nightly build. During the update, you would get fatal error saying class `Joomla\Plugin\Actionlog\Joomla\Extension\Joomla` not found (something like that, I don't remember exactly the error message)
3. Install Joomla 4.1 again
4. Try to update to the update package generated for this PR. You can download the package from this link https://ci.joomla.org/artifacts/joomla/joomla-cms/4.2-dev/37990/downloads/54926/Joomla_4.2.0-beta1-dev+pr.37990-Development-Update_Package.zip . Make sure upgrade works properly

### Actual result BEFORE applying this Pull Request
Fatal error while upgrading to Joomla 4.2

### Expected result AFTER applying this Pull Request
Upgrade to Joomla 4.2 works well.

joomdonation committed:
Fix fatal error when update to Joomla 4.2

ChristineWk commented:
@joomdonation 
Confirm Actual Result. Updated from 4.1.4 to Nightly Build 4.2.0-beta1-dev.

`An error has occurred.
0 Class 'Joomla\Plugin\Actionlog\Joomla\Extension\Joomla' not found `

But now, Frontend and Backend are corrupted.
Can I solve it via phpMyAdmin somewhere (special Plugin Table)?

joomdonation commented:
@ChristineWk You can delete the file administrator/cache/autoload_psr4.php manually to get the site running again. However, to test the PR, you will need to setup another Joomla 4.1 website. The purpose is confirming upgrading to new package won't get fatal error anymore

ChristineWk commented:
> @ChristineWk You can delete the file administrator/cache/autoload_psr4.php manually to get the site running again. However, to test the PR, you will need to setup another Joomla 4.1 website. The purpose is confirming upgrading to new package won't get fatal error anymore

OK, Done. Thanks for the hint :-)
It shows now: 4.2.0-beta1-dev and: Database: 4.1.3-2022-04-08:  After this: DB updated.
Another 4.1 I don't know yet.

brianteeman commented:
From what I remember that cache file is supposed to be deleted and rebuiult during the update process

brianteeman commented:
found it https://github.com/joomla/joomla-cms/pull/36094

joomdonation commented:
@brianteeman That PR only solved the issue when new extensions are added during update process (like Schedule extension which we added in 4.1). For 4.2, we converted some plugins to new structure, so the namespace map must be re-generated earlier than that, otherwise, after the extraction process, we will get fatal error immediately and the upgrade could not be completed.

richard67 commented:
I have tested this item :white_check_mark: successfully on 7aaebc4df056cff1a632f69ba6690be94f2bfa1b<br /><br />Hint for other testers: For updating to the 4.2-dev nightly build without the patch (step 2 in the instructions), I had to download the update package from here https://developer.joomla.org/nightly-builds.html and use "Upload & Update" because the update site XML for the 4.2-dev nightlies contain a wrong download URL. See https://github.com/joomla/update.joomla.org/pull/250 .<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

richard67 commented on administrator/components/com_joomlaupdate/finalisation.php:
```diff
@@ -75,6 +75,17 @@ function finalizeUpdate(string $siteRoot, string $restorePath): void
+			 */
+			$namespaceMapFile = JPATH_ROOT . '/administrator/cache/autoload_psr4.php';
+
+			if (\Joomla\CMS\Filesystem\File::exists($namespaceMapFile))
+			{
+				\Joomla\CMS\Filesystem\File::delete($namespaceMapFile);
+			}
```
@joomdonation Will this also work in case if the mocked classed are used which are defined a bit below in this file with their namespace? Or shouldn't we better use the native PHP methods here? I have no idea, am just asking.

joomdonation replied:
@richard67 I guess PHP compile the code before it is executed, so that should not be any problem (you and me both tested and it worked). We should try to use library method when it is available than use the native php code (as you can see, the delete method not only delete file but also handle opcache). So the code is fine as it is.

laoneo replied:
Please use the classes from the filesystem package and not from the CMS.

richard67 replied:
@laoneo I think you are wrong here because the CMS classes are the ones mocked a bit below in that file. And script.php also uses these, see here: https://github.com/joomla/joomla-cms/blob/4.2-dev/administrator/components/com_admin/script.php#L14-L15

If you want to change that, you have to change it everywhere in the CMS update process, i.e. also in script.php, and that is out of scope of this PR here.

richard67 replied:
> @richard67 I guess PHP compile the code before it is executed, so that should not be any problem (you and me both tested and it worked). We should try to use library method when it is available than use the native php code (as you can see, the delete method not only delete file but also handle opcache). So the code is fine as it is.

@joomdonation Thanks, you are right.

laoneo replied:
@richard67 you are right. No I wonder why it is like that. But yes not part of this pr.

joomdonation replied:
@richard67 @laoneo The actual reason is the code in this file is not running in Joomla environment, it's running as a standalone PHP script, so the Joomla library code is not available to use here. That's the reason we see the code defined in that file, and we could not use the Joomla framework filesystem package here.

toivo commented:
I have tested this item :white_check_mark: successfully on 7aaebc4df056cff1a632f69ba6690be94f2bfa1b<br /><br />Confirmed the error message and tested successfully in Wampserver 3.2.9 and PHP 8.0.15 using the update package created for this PR.

```
#
#<?php die('Forbidden.'); ?>
#Date: 2022-06-06 03:43:55 UTC
#Software: Joomla! 4.1.4 Stable [ Kuamini ] 24-May-2022 16:00 GMT

#Fields: datetime	priority clientip	category	message
2022-06-06T03:43:55+00:00	INFO 127.0.0.1	update	Starting installation of new version.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Finalising installation.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Start of SQL updates.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	The current database version (schema) is 4.1.3-2022-04-08.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: CREATE TABLE IF NOT EXISTS `#__user_mfa` (   `id` int NOT NULL AUTO_INCREMENT,  .
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: DELETE FROM `#__postinstall_messages` WHERE `condition_file` = 'site://plugins/t.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, .
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: UPDATE `#__extensions` AS `a` 	INNER JOIN `#__extensions` AS `b` on `a`.`element.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: DELETE FROM `#__extensions` WHERE `type` = 'plugin' AND `folder` = 'twofactoraut.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: INSERT IGNORE INTO `#__postinstall_messages` (`extension_id`, `title_key`, `desc.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Ran query from file 4.2.0-2022-05-15. Query text: INSERT IGNORE INTO `#__mail_templates` (`template_id`, `extension`, `language`, .
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	End of SQL updates.
2022-06-06T03:44:15+00:00	INFO 127.0.0.1	update	Deleting removed files and folders.
2022-06-06T03:44:19+00:00	INFO 127.0.0.1	update	Cleaning up after installation.
2022-06-06T03:44:19+00:00	INFO 127.0.0.1	update	Update to version 4.2.0-beta1-dev+pr.37990 is complete.
```

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

roland-d committed:
Merge branch '4.2-dev' into fix_fatal_error_on_upgrade_to_4.2

roland-d commented:
Thanks everybody

PhilETaylor commented:
Cross posting https://github.com/joomla/joomla-cms/issues/38474 for visibility - Joomal 4.2.0 Upgrade bricks sites...

shortyloz commented:
I have a related issue maybe?

![screen shot 2022-08-16 at 21 12 58](https://issues.joomla.org/uploads/1/5ed6a0664350f1265d733fef5570aed1.png)

I have removed the file suggested above "@ChristineWk You can delete the file administrator/cache/autoload_psr4.php manually to get the site running again. However, to test the PR, you will need to setup another Joomla 4.1 website. The purpose is confirming upgrading to new package won't get fatal error anymore"  But that has not solved my issue.

Any suggestions : Live site is at this URL (I also stupidly did not do a backup!)
https://www.cheshuntrpc.co.uk/

Thanks
Numpty Mark from UK
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

PhilETaylor commented:
set debug mode on and error reporting to development in configuration.php to get a stack trace - this is a PHP 8 issue and unrelated to autoload_psr4

PhilETaylor commented:
Community Support is best obtained through the https://forum.joomla.org and not here.

You can edit /configuration.php with a editor and change `$debug = false` to `$debug = true` and `$error_reporting = 'none'` to `$error_reporting = 'development'` and then save the file back to your site.

shortyloz commented:
Sorry, I didn't know I was in the wrong place.  I will do that as suggested and also look there. :-)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

rkostanyan commented:
Dears, I have new problem, please help
Class "Joomla\Plugin\Authentication\Joomla\Extension\Joomla" not found<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

joomdonation commented:
@rkostanyan Try to delete the file administrator/cache/autoload_psr4.php from your site. That should address the issue.

rkostanyan commented:
Problem is not resolved:(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37990">issues.joomla.org/tracker/joomla-cms/37990</a>.</sub>

joomdonation commented:
@rkostanyan If so, I don't have any other idea. You might need to have a developer to look at your site to see what's actual reason and get it sorted. Without accessing to your site, it's hard to know the root reason of the error
